lib/utils/authentication.js

```diff
@@ -44,23 +45,114 @@ module.exports.register = function(server, options, next) {
      * Even though this is a synchronous function, it returns a promise
      * This will allow for an easy transition to async key generators
      * @param {string} username - unencrypted username
-     * @return {Bluebird} resolves to a credentials object
+     * @return {Promise} resolves to a credentials object
      *   where credentials is an object with the following properties:
-     *   `id, key, algorithm, issueTime, expireTime`
+     *   `id, key, algorithm, issueTime, expireTime, studyRoles`
      */
     authUtils.generateHawkCredentials = function(username) {
-        const credentials = {
-            username: username,
-            id: uuid.v4(),
-            key: uuid.v4(),
-            algorithm: config.algorithm,
-            issueTime: +new Date(),
-            expireTime: +new Date() + config.hawkKeyLifespan
-        };
-        return Bluebird.resolve(credentials);
+
+        return refreshCurrentStudyRoles(username)
+            .then((roleList) => {
+                const credentials = {
+                    username: username,
+                    id: uuid.v4(),
+                    key: uuid.v4(),
+                    algorithm: config.algorithm,
+                    issueTime: + new Date(),
+                    expireTime: + new Date() + config.hawkKeyLifespan,
+                    studyRoles: roleList
+                };
+
+                return credentials;
+            });
     };
 
     /**
+     * Retrieves the current roles for a user, revokes old roles, and
+     * adds the new roles
+     * @param {string} username - the username to retrieve roles
+     * @return {object} roles - the current roles
+     */
+    function refreshCurrentStudyRoles(username) {
+        const encryptedUsername = authUtils.encryptUsername(username);
+
+        const UserRole = server.plugins.bookshelf.model('UserStudyRole');
+    
+        return new UserRole()
+            .where({username: encryptedUsername})
+            .fetchAll({withRelated: 'role'})
+            .then((roleCollection) => {
```

Comment:
In fact, it would be best to not use partials for variables that are defined asynchronously. Might be good to have a group discussion about the best way to handle this in the future.